hpx/parallel/algorithms/sort.hpp

```diff
@@ -201,19 +202,22 @@ namespace hpx { namespace parallel { inline namespace v1
                 result = execution::async_execute(policy.executor(),
                         &sort_thread<ExPolicy, RandomIt, Compare>,
                         std::ref(policy), first, last, comp);
+
+                return result;
             }
             catch (...) {
-                return detail::handle_exception<ExPolicy, RandomIt>::call(
-                    std::current_exception());
-            }
+                try {
+                    util::detail::handle_local_exceptions<ExPolicy>::call(
+                        std::current_exception());
 
-            if (result.has_exception())
-            {
-                return detail::handle_exception<ExPolicy, RandomIt>::call(
-                    std::move(result));
+                    // Not reachable.
+                    HPX_ASSERT(false);
+                }
+                catch (...) {
+                    return hpx::make_exceptional_future<RandomIt>(
+                        std::current_exception());
+                }
```

Comment:
Hmmm, do we really have to use nested try/catch blocks here? `handle_local_exceptions` only rethrows the current exception as an exception list. Couldn't we directly call `make_exceptional_future` with a exception list constructed from the current exception?

Reply:
@hkaiser ```handle_local_exceptions()`` handles an exception differently when ```std::current_exception()``` means ```std::bad_alloc```. In that case, ```handle_local_exceptions()``` just rethrows ```std::bad_alloc``` without wrapping into ```exception_list```.
We need to handle the exception differently according to what is the exception which we handles. And for that, I used ```handle_local_exception()```.
And nested catch (Line216-219) is needed for wrapping an exception which is threw properly from ```handle_local_exception()```.

Reply:
@taeguk yes, I understand. Still, there might be a better way.

Reply:
@hkaiser Aha, okay i got it.
https://github.com/taeguk/hpx/blob/9d72e4de54495987597bc2e6bbe609bc8125a01c/hpx/parallel/algorithms/sort.hpp#L249-L260
If we move that exception handling from Line208 to above lines, we can avoid nested try/catch blocks with just calling ```detail::handle_exception()```.

Reply:
For example, https://github.com/taeguk/hpx/blob/9d72e4de54495987597bc2e6bbe609bc8125a01c/hpx/parallel/algorithms/partition.hpp#L963-L975